src/sockets/chatSpamFilter.ts

```diff
@@ -0,0 +1,38 @@
+export class ChatSpamFilter {
+  history: { [username: string]: number[] };
+
+  seconds: number;
+
+  constructor() {
+    this.history = {};
+    this.seconds = 0;
+  }
+
+  // Returns true if user can chat. Else false.
+  chatRequest(username: string): boolean {
+    if (!(username in this.history)) {
+      this.history[username] = [this.seconds];
+    } else {
+      const userHistory = this.history[username];
+      userHistory.push(this.seconds);
+      if (userHistory.length > 4) {
+        if (userHistory[4] - userHistory[0] < 5) {
+          userHistory.shift();
+          return false;
+        }
+        userHistory.shift();
+      }
+    }
+
+    return true;
+  }
+
+  // Users of this class must call this every second!
+  tick() {
+    this.seconds++;
+  }
+
+  getSeconds(): number {
+    return this.seconds;
+  }
+}
```

Comment:
Let's also add a method to remove a user when they disconnect, so we don't grow the chat object too much. 
Look into `socket.on('disconnect')` in socket.ts to hook it in.